### ISSUE TITLE
Nvidia optimus not working

### ISSUE THREAD

**Issue by LordSuricato** [LizardByte/Sunshine#994]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

I have a Dell XPS 7590 with an Intel UHD 630 and an NVIDIA GTX 1650. My Dell uses Nvidia Optimus and Sunshine will not boot if I set `adapter_name = NVIDIA GeForce GTX 1650`. Unconfigured it uses the UHD 630 and works correctly but the encoding performance is more limited than using the GTX 1650.

### Expected Behavior

I would expect to put `adapter_name = NVIDIA GeForce GTX 1650` and use that hardware for Sunshine encoding.

### Host Operating System

Windows

### Operating System Version

Windows 11 

### Architecture

64 bit

### Sunshine commit or version

0.18.3 and nightly

### Package

Windows - installer

### GPU Type

Nvidia

### GPU Model

GeForce GTX 1650

### GPU Driver/Mesa Version

latest

### Config

```Shell
key_rightalt_to_key_win = disabled
gamepad = x360
upnp = disabled
dwmflush = enabled
min_log_level = 2
origin_pin_allowed = pc
origin_web_ui_allowed = lan
hevc_mode = 0
nv_preset = p4
nv_tune = ull
nv_coder = auto
nv_rc = cbr
qsv_preset = medium
qsv_coder = auto
amd_coder = auto
amd_quality = balanced
amd_rc = vbr_latency
vt_coder = auto
vt_software = auto
vt_realtime = enabled
fps = [10,30,60,90,120]
resolutions = [
    352x240,
    480x360,
    858x480,
    1280x720,
    1920x1080,
    2560x1080,
    3440x1440,
    1920x1200,
    3860x2160,
    3840x1600
]
adapter_name = NVIDIA GeForce GTX 1650
```

### Apps

```json
Desktop
```

### Relevant log output

```shell
[vt_coder] -- [auto]
[amd_rc] -- [vbr_latency]
[vt_realtime] -- [enabled]
[key_rightalt_to_key_win] -- [disabled]
[resolutions] -- [[
    352x240,
    480x360,
    858x480,
    1280x720,
    1920x1080,
    2560x1080,
    3440x1440,
    1920x1200,
    3860x2160,
    3840x1600
]]
[origin_pin_allowed] -- [pc]
[qsv_coder] -- [auto]
[gamepad] -- [x360]
[adapter_name] -- [NVIDIA GeForce GTX 1650]
[amd_coder] -- [auto]
[nv_preset] -- [p4]
[upnp] -- [disabled]
[nv_coder] -- [auto]
[dwmflush] -- [enabled]
[min_log_level] -- [2]
[vt_software] -- [auto]
[amd_quality] -- [balanced]
[origin_web_ui_allowed] -- [lan]
[nv_rc] -- [cbr]
[hevc_mode] -- [0]
[nv_tune] -- [ull]
[fps] -- [[10,30,60,90,120]]
[qsv_preset] -- [medium]
[2023:03:03:00:47:29]: Info: Compiling shaders...
[2023:03:03:00:47:30]: Info: Compiled shaders
[2023:03:03:00:47:30]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2023:03:03:00:47:30]: Info: Trying encoder [nvenc]
[2023:03:03:00:47:30]: Info: ddprobe.exe [1] [] returned: 0x00000000
[2023:03:03:00:47:30]: Info: Set GPU preference: 1
[2023:03:03:00:47:30]: Error: Failed to locate an output device
[2023:03:03:00:47:30]: Error: Failed to locate an output device
[2023:03:03:00:47:30]: Error: Failed to locate an output device
[2023:03:03:00:47:30]: Error: Failed to locate an output device
[2023:03:03:00:47:30]: Error: Failed to locate an output device
[2023:03:03:00:47:31]: Error: Failed to locate an output device
[2023:03:03:00:47:31]: Error: Failed to locate an output device
[2023:03:03:00:47:31]: Error: Failed to locate an output device
[2023:03:03:00:47:31]: Info: Encoder [nvenc] failed
[2023:03:03:00:47:31]: Info: Trying encoder [quicksync]
[2023:03:03:00:47:31]: Error: Failed to locate an output device
[2023:03:03:00:47:31]: Error: Failed to locate an output device
[2023:03:03:00:47:32]: Error: Failed to locate an output device
[2023:03:03:00:47:32]: Error: Failed to locate an output device
[2023:03:03:00:47:32]: Error: Failed to locate an output device
[2023:03:03:00:47:32]: Error: Failed to locate an output device
[2023:03:03:00:47:32]: Error: Failed to locate an output device
[2023:03:03:00:47:33]: Error: Failed to locate an output device
[2023:03:03:00:47:33]: Info: Encoder [quicksync] failed
[2023:03:03:00:47:33]: Info: Trying encoder [amdvce]
[2023:03:03:00:47:33]: Error: Failed to locate an output device
[2023:03:03:00:47:33]: Error: Failed to locate an output device
[2023:03:03:00:47:33]: Error: Failed to locate an output device
[2023:03:03:00:47:34]: Error: Failed to locate an output device
[2023:03:03:00:47:34]: Error: Failed to locate an output device
[2023:03:03:00:47:34]: Error: Failed to locate an output device
[2023:03:03:00:47:34]: Error: Failed to locate an output device
[2023:03:03:00:47:34]: Error: Failed to locate an output device
[2023:03:03:00:47:35]: Info: Encoder [amdvce] failed
[2023:03:03:00:47:35]: Info: Trying encoder [software]
[2023:03:03:00:47:35]: Error: Failed to locate an output device
[2023:03:03:00:47:35]: Error: Failed to locate an output device
[2023:03:03:00:47:35]: Error: Failed to locate an output device
[2023:03:03:00:47:35]: Error: Failed to locate an output device
[2023:03:03:00:47:35]: Error: Failed to locate an output device
[2023:03:03:00:47:36]: Error: Failed to locate an output device
[2023:03:03:00:47:36]: Error: Failed to locate an output device
[2023:03:03:00:47:36]: Error: Failed to locate an output device
[2023:03:03:00:47:36]: Info: Encoder [software] failed
[2023:03:03:00:47:36]: Fatal: Couldn't find any working encoder
```

----

```
====== ADAPTER =====
Device Name      : NVIDIA GeForce GTX 1650
Device Vendor ID : 0x000010DE
Device Device ID : 0x00001F91
Device Video Mem : 3951 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 16243 MiB

    ====== OUTPUT ======
    Output Name       : \\.\DISPLAY1
    AttachedToDesktop : yes
    Resolution        : 1536x864

====== ADAPTER =====
Device Name      : Intel(R) UHD Graphics 630
Device Vendor ID : 0x00008086
Device Device ID : 0x00003E9B
Device Video Mem : 128 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 16243 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : Microsoft Basic Render Driver
Device Vendor ID : 0x00001414
Device Device ID : 0x0000008C
Device Video Mem : 0 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 16243 MiB

    ====== OUTPUT ======
```
```

**Comment by cgutman** [comment]
This is expected behavior. The Desktop Duplication API used for capture is only avilable on the GPU used for display output.

**Comment by Nonary** [comment]
The only way to force GPU encoding for NVIDIA Optimus machines is basically the following.

1. Hope your laptop has an MUX switch https://www.theworldsbestandworst.com/2021/12/how-to-turn-off-optimus-on-your-gaming-laptop/
2. In the event you don't have an MUX switch, attach an HDMI/Display Port dummy plug as that will attach directly to GPU.


**Note** You can't use USB docks with an HDMI/DP port because they do not attach directly to GPU in that scenario.


3. Live with it and stream with iGPU, complain at your laptop manufacturer for not providing MUX switch.

**Comment by LordSuricato** [comment]
> La única forma de forzar la codificación GPU para las máquinas NVIDIA Optimus es básicamente la siguiente.
> 
> 1. Espero que su computadora portátil tenga un interruptor MUX https://www.theworldsbestandworst.com/2021/12/how-to-turn-off-optimus-on-your-gaming-laptop/
> 2. En caso de que no tenga un conmutador MUX, conecte un conector falso HDMI/Display Port, ya que se conectará directamente a la GPU.
> 
> **Nota** No puede usar bases USB con un puerto HDMI/DP porque no se conectan directamente a la GPU en ese escenario.
> 
> 3. Viva con él y transmita con iGPU, quéjese con el fabricante de su computadora portátil por no proporcionar el interruptor MUX.

1- Dell XPS 7590 does not have Mux Switch. I think no Dell XPS.

2- I also tried it with an HDMI dummy, a new screen appears in output but the result is the same.


3 - I wrote to Dell but I don't think I have any results 


I was just testing performance anyway. My goal is to buy a tower computer with an RTX 4000 series. On the other hand, I have an Intel NUC i7 at home that I use as an application server, it has an Intel UHD 630 and it works perfect for working remotely developing. 

I close the issue because it is not an issue, it is the expected behavior with the HW of my laptop. 

Thank you very much for this development, I usually donate to projects that solve my problems and this is one of them. Sunshine allows me to work with Galaxy TAB s8 Ultra. Wherever and however you want, it also works to play video games.


TLDR: Laptops that use Nvidia Optimus to save battery life that do not have MUX Switch are not supported and even if they are, the performance may not be the best.

Thanks

**Comment by Originalimoc** [comment]
@cgutman Steam Link proves significantly better performance so the worry about not good performance is not correct.

**Comment by cgutman** [comment]
Steam Link doesn't use the Desktop Duplication API last I checked, so reasoning about its performance vs a hypothetical Sunshine solution that uses DDAPI+copyback isn't really valid.

**Comment by cgutman** [comment]
> Why "hypothetical"? Indeed, its capture method is hybrid depending on the program running and/or Steam Link client requested game vs what's actually on screen sometime can use NVFBC for lowest latency possible, its capture method switching is dynamic, so more complex than sunshine right now. 

NvFBC is [deprecated in favor of DDAPI on Windows](https://developer.download.nvidia.com/designworks/capture-sdk/docs/NVFBC_Win10_Deprecation_Tech_Bulletin.pdf) (and license-encumbered by Nvidia), so we won't/can't adopt it in Sunshine.

> But since Sunshine advertise self as gaming solution, and assuming DDAPI can be replaced, frame if you want to do SW encoding are in RAM anyway, the cost of copying to dGPU is relatively cheap compare to encoding on CPU especially above 1080P60 + HEVC which only Zen2+ performance can reasonably playable. Old 4/6 core intel you lost about 2 cores doing this encoding, only room for some 2D and small 3D game then. Also much lower latency I've tested too. This can on the todo list.

There's no reason to replace DDAPI. It is the recommended API for high-performance display capture, even supporting things Steam Link cannot such as HDR.

We already support copy-back for CPU encoding with DDAPI today. This thread is referring to _hardware_ encoding on the iGPU, which is already supported today for both AMD and Intel iGPUs. The performance concern is directly doing encoding on the iGPU where the captured frame already resides vs. copying the frame from iGPU to the CPU and back out to the dGPU for encoding. I'm unconvinced the latter is a win for performance or power efficiency.
